# LSTM

In [1]:
'''Trains an LSTM model on the IMDB sentiment classification task.
The dataset is actually too small for LSTM to be of any advantage
compared to simpler, much faster methods such as TF-IDF + LogReg.
# Notes
- RNNs are tricky. Choice of batch size is important,
choice of loss and optimizer is critical, etc.
Some configurations won't converge.
- LSTM loss decrease patterns during training can be quite different
from what you see with CNNs/MLPs/etc.
'''
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import LSTM
from keras.datasets import imdb

max_features = 20000
# cut texts after this number of words (among top max_features most common words)
maxlen = 80
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(LSTM(128, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(1, activation='sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=5,
          validation_data=(x_test, y_test))
score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

Using TensorFlow backend.


Loading data...
17465344/17464789 [==============================] - 3s 0us/step
25000 train sequences
25000 test sequences
Pad sequences (samples x time)
x_train shape: (25000, 80)
x_test shape: (25000, 80)
Build model...
Train...
Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 172s 7ms/step - loss: 0.4588 - acc: 0.7830 - val_loss: 0.4140 - val_acc: 0.8108
Epoch 2/5
25000/25000 [==============================] - 166s 7ms/step - loss: 0.3015 - acc: 0.8780 - val_loss: 0.3732 - val_acc: 0.8367
Epoch 3/5
25000/25000 [==============================] - 165s 7ms/step - loss: 0.2190 - acc: 0.9153 - val_loss: 0.4173 - val_acc: 0.8231
Epoch 4/5
25000/25000 [==============================] - 170s 7ms/step - loss: 0.1587 - acc: 0.9396 - val_loss: 0.4451 - val_acc: 0.8198
Epoch 5/5
25000/25000 [==============================] - 43s 2ms/step
Test score: 0.5874459129095078
Test accuracy: 0.81792


# SimpleRNN

In [2]:
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import SimpleRNN
from keras.datasets import imdb

max_features = 20000
# cut texts after this number of words (among top max_features most common words)
maxlen = 80
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(SimpleRNN(128, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(1, activation='sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=5,
          validation_data=(x_test, y_test))
score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

Loading data...
25000 train sequences
25000 test sequences
Pad sequences (samples x time)
x_train shape: (25000, 80)
x_test shape: (25000, 80)
Build model...
Train...
Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 44s 2ms/step - loss: 0.6913 - acc: 0.5342 - val_loss: 0.6641 - val_acc: 0.5936
Epoch 2/5
25000/25000 [==============================] - 45s 2ms/step - loss: 0.6275 - acc: 0.6368 - val_loss: 0.6167 - val_acc: 0.6400
Epoch 3/5
25000/25000 [==============================] - 44s 2ms/step - loss: 0.5170 - acc: 0.7450 - val_loss: 0.4985 - val_acc: 0.7691
Epoch 4/5
25000/25000 [==============================] - 44s 2ms/step - loss: 0.4503 - acc: 0.7967 - val_loss: 0.4956 - val_acc: 0.7791
Epoch 5/5
25000/25000 [==============================] - 11s 445us/step
Test score: 0.5208023117828369
Test accuracy: 0.75712


# GRU

In [4]:
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import GRU
from keras.datasets import imdb

max_features = 20000
# cut texts after this number of words (among top max_features most common words)
maxlen = 80
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(GRU(128, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(1, activation='sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=5,
          validation_data=(x_test, y_test))
score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

Loading data...
25000 train sequences
25000 test sequences
Pad sequences (samples x time)
x_train shape: (25000, 80)
x_test shape: (25000, 80)
Build model...
Train...
Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 128s 5ms/step - loss: 0.4738 - acc: 0.7709 - val_loss: 0.3835 - val_acc: 0.8274
Epoch 2/5
25000/25000 [==============================] - 129s 5ms/step - loss: 0.2880 - acc: 0.8835 - val_loss: 0.3771 - val_acc: 0.8428
Epoch 3/5
25000/25000 [==============================] - 130s 5ms/step - loss: 0.1789 - acc: 0.9330 - val_loss: 0.4421 - val_acc: 0.8338
Epoch 4/5
25000/25000 [==============================] - 130s 5ms/step - loss: 0.1063 - acc: 0.9623 - val_loss: 0.5183 - val_acc: 0.8179
Epoch 5/5
25000/25000 [==============================] - 33s 1ms/step
Test score: 0.6337903826999665
Test accuracy: 0.8212


# ConvLSTM2D

In [5]:
""" This script demonstrates the use of a convolutional LSTM network.
This network is used to predict the next frame of an artificially
generated movie which contains moving squares.
"""
from keras.models import Sequential
from keras.layers.convolutional import Conv3D
from keras.layers.convolutional_recurrent import ConvLSTM2D
from keras.layers.normalization import BatchNormalization
import numpy as np
import pylab as plt

# We create a layer which take as input movies of shape
# (n_frames, width, height, channels) and returns a movie
# of identical shape.

seq = Sequential()
seq.add(ConvLSTM2D(filters=40, kernel_size=(3, 3),
                   input_shape=(None, 40, 40, 1),
                   padding='same', return_sequences=True))
seq.add(BatchNormalization())

seq.add(ConvLSTM2D(filters=40, kernel_size=(3, 3),
                   padding='same', return_sequences=True))
seq.add(BatchNormalization())

seq.add(ConvLSTM2D(filters=40, kernel_size=(3, 3),
                   padding='same', return_sequences=True))
seq.add(BatchNormalization())

seq.add(ConvLSTM2D(filters=40, kernel_size=(3, 3),
                   padding='same', return_sequences=True))
seq.add(BatchNormalization())

seq.add(Conv3D(filters=1, kernel_size=(3, 3, 3),
               activation='sigmoid',
               padding='same', data_format='channels_last'))
seq.compile(loss='binary_crossentropy', optimizer='adadelta')


# Artificial data generation:
# Generate movies with 3 to 7 moving squares inside.
# The squares are of shape 1x1 or 2x2 pixels,
# which move linearly over time.
# For convenience we first create movies with bigger width and height (80x80)
# and at the end we select a 40x40 window.

def generate_movies(n_samples=1200, n_frames=15):
    row = 80
    col = 80
    noisy_movies = np.zeros((n_samples, n_frames, row, col, 1), dtype=np.float)
    shifted_movies = np.zeros((n_samples, n_frames, row, col, 1),
                              dtype=np.float)

    for i in range(n_samples):
        # Add 3 to 7 moving squares
        n = np.random.randint(3, 8)

        for j in range(n):
            # Initial position
            xstart = np.random.randint(20, 60)
            ystart = np.random.randint(20, 60)
            # Direction of motion
            directionx = np.random.randint(0, 3) - 1
            directiony = np.random.randint(0, 3) - 1

            # Size of the square
            w = np.random.randint(2, 4)

            for t in range(n_frames):
                x_shift = xstart + directionx * t
                y_shift = ystart + directiony * t
                noisy_movies[i, t, x_shift - w: x_shift + w,
                             y_shift - w: y_shift + w, 0] += 1

                # Make it more robust by adding noise.
                # The idea is that if during inference,
                # the value of the pixel is not exactly one,
                # we need to train the network to be robust and still
                # consider it as a pixel belonging to a square.
                if np.random.randint(0, 2):
                    noise_f = (-1)**np.random.randint(0, 2)
                    noisy_movies[i, t,
                                 x_shift - w - 1: x_shift + w + 1,
                                 y_shift - w - 1: y_shift + w + 1,
                                 0] += noise_f * 0.1

                # Shift the ground truth by 1
                x_shift = xstart + directionx * (t + 1)
                y_shift = ystart + directiony * (t + 1)
                shifted_movies[i, t, x_shift - w: x_shift + w,
                               y_shift - w: y_shift + w, 0] += 1

    # Cut to a 40x40 window
    noisy_movies = noisy_movies[::, ::, 20:60, 20:60, ::]
    shifted_movies = shifted_movies[::, ::, 20:60, 20:60, ::]
    noisy_movies[noisy_movies >= 1] = 1
    shifted_movies[shifted_movies >= 1] = 1
    return noisy_movies, shifted_movies

# Train the network
noisy_movies, shifted_movies = generate_movies(n_samples=1200)
seq.fit(noisy_movies[:1000], shifted_movies[:1000], batch_size=10,
        epochs=5, validation_split=0.05)

# Testing the network on one movie
# feed it with the first 7 positions and then
# predict the new positions
which = 1004
track = noisy_movies[which][:7, ::, ::, ::]

for j in range(16):
    new_pos = seq.predict(track[np.newaxis, ::, ::, ::, ::])
    new = new_pos[::, -1, ::, ::, ::]
    track = np.concatenate((track, new), axis=0)


# And then compare the predictions
# to the ground truth
track2 = noisy_movies[which][::, ::, ::, ::]
for i in range(15):
    fig = plt.figure(figsize=(10, 5))

    ax = fig.add_subplot(121)

    if i >= 7:
        ax.text(1, 3, 'Predictions !', fontsize=20, color='w')
    else:
        ax.text(1, 3, 'Initial trajectory', fontsize=20)

    toplot = track[i, ::, ::, 0]

    plt.imshow(toplot)
    ax = fig.add_subplot(122)
    plt.text(1, 3, 'Ground truth', fontsize=20)

    toplot = track2[i, ::, ::, 0]
    if i >= 2:
        toplot = shifted_movies[which][i - 1, ::, ::, 0]

    plt.imshow(toplot)
plt.savefig('%i_animate.png' % (i + 1))

Train on 950 samples, validate on 50 samples
Epoch 1/5
950/950 [==============================] - 41s 43ms/step - loss: 0.2708 - val_loss: 0.0641
Epoch 2/5
950/950 [==============================] - 33s 35ms/step - loss: 0.0269 - val_loss: 0.0123
Epoch 3/5
950/950 [==============================] - 33s 35ms/step - loss: 0.0043 - val_loss: 0.0034
Epoch 4/5
950/950 [==============================] - 33s 35ms/step - loss: 0.0013 - val_loss: 9.4860e-04
Epoch 5/5
950/950 [==============================] - 33s 35ms/step - loss: 7.5464e-04 - val_loss: 5.8110e-04


# SimpleRNNCell

In [6]:
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import RNN, SimpleRNNCell
from keras.datasets import imdb

max_features = 20000
# cut texts after this number of words (among top max_features most common words)
maxlen = 80
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('New stacked LSTM: 30s/epoch on CPU (15pct faster)')
cells = [
    SimpleRNNCell(128, dropout=0.2, recurrent_dropout=0.2),
]

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(RNN(cells))
model.add(Dense(1, activation='sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=5,
          validation_data=(x_test, y_test))
score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

Loading data...
25000 train sequences
25000 test sequences
Pad sequences (samples x time)
x_train shape: (25000, 80)
x_test shape: (25000, 80)
New stacked LSTM: 30s/epoch on CPU (15pct faster)
Build model...
Train...
Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 45s 2ms/step - loss: 0.6876 - acc: 0.5409 - val_loss: 0.6708 - val_acc: 0.5528
Epoch 2/5
25000/25000 [==============================] - 44s 2ms/step - loss: 0.6223 - acc: 0.6451 - val_loss: 0.6315 - val_acc: 0.6222
Epoch 3/5
25000/25000 [==============================] - 44s 2ms/step - loss: 0.5443 - acc: 0.7186 - val_loss: 0.6657 - val_acc: 0.6307
Epoch 4/5
25000/25000 [==============================] - 44s 2ms/step - loss: 0.4727 - acc: 0.7679 - val_loss: 0.6088 - val_acc: 0.6953
Epoch 5/5
25000/25000 [==============================] - 11s 450us/step
Test score: 0.6421183574485779
Test accuracy: 0.7148


# 各処理の結果比較

|実行時間  |Test score  |Test accuracy  |
|---|---|---|---|
|LSTM  |2  |3  |
|4  |5  |6  |

# GRUCell

In [7]:
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import RNN, GRUCell
from keras.datasets import imdb

max_features = 20000
# cut texts after this number of words (among top max_features most common words)
maxlen = 80
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('New stacked LSTM: 30s/epoch on CPU (15pct faster)')
cells = [
    GRUCell(128, dropout=0.2, recurrent_dropout=0.2),
]

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(RNN(cells))
model.add(Dense(1, activation='sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=5,
          validation_data=(x_test, y_test))
score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

Loading data...
25000 train sequences
25000 test sequences
Pad sequences (samples x time)
x_train shape: (25000, 80)
x_test shape: (25000, 80)
New stacked LSTM: 30s/epoch on CPU (15pct faster)
Build model...
Train...
Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 134s 5ms/step - loss: 0.4775 - acc: 0.7691 - val_loss: 0.4314 - val_acc: 0.8006
Epoch 2/5
25000/25000 [==============================] - 142s 6ms/step - loss: 0.2909 - acc: 0.8812 - val_loss: 0.3586 - val_acc: 0.8465
Epoch 3/5
25000/25000 [==============================] - 129s 5ms/step - loss: 0.1826 - acc: 0.9302 - val_loss: 0.4548 - val_acc: 0.8247
Epoch 4/5
25000/25000 [==============================] - 131s 5ms/step - loss: 0.1088 - acc: 0.9615 - val_loss: 0.5372 - val_acc: 0.8148
Epoch 5/5
25000/25000 [==============================] - 31s 1ms/step
Test score: 0.6497126778757573
Test accuracy: 0.8196


# LSTMCell

In [8]:
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import RNN, LSTMCell
from keras.datasets import imdb

max_features = 20000
# cut texts after this number of words (among top max_features most common words)
maxlen = 80
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('New stacked LSTM: 30s/epoch on CPU (15pct faster)')
cells = [
    LSTMCell(128, dropout=0.2, recurrent_dropout=0.2),
]

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(RNN(cells))
model.add(Dense(1, activation='sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=5,
          validation_data=(x_test, y_test))
score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

Loading data...
25000 train sequences
25000 test sequences
Pad sequences (samples x time)
x_train shape: (25000, 80)
x_test shape: (25000, 80)
New stacked LSTM: 30s/epoch on CPU (15pct faster)
Build model...
Train...
Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 178s 7ms/step - loss: 0.4573 - acc: 0.7854 - val_loss: 0.3834 - val_acc: 0.8310
Epoch 2/5
25000/25000 [==============================] - 177s 7ms/step - loss: 0.3036 - acc: 0.8758 - val_loss: 0.3730 - val_acc: 0.8370
Epoch 3/5
25000/25000 [==============================] - 170s 7ms/step - loss: 0.2169 - acc: 0.9165 - val_loss: 0.5098 - val_acc: 0.7771
Epoch 4/5
25000/25000 [==============================] - 169s 7ms/step - loss: 0.1583 - acc: 0.9400 - val_loss: 0.4649 - val_acc: 0.8264
Epoch 5/5
25000/25000 [==============================] - 42s 2ms/step
Test score: 0.5684661542415619
Test accuracy: 0.81792


# StackedRNNCells

In [9]:
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import RNN, LSTMCell
from keras.datasets import imdb

max_features = 20000
# cut texts after this number of words (among top max_features most common words)
maxlen = 80
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('New stacked LSTM: 30s/epoch on CPU (15pct faster)')
cells = [
    LSTMCell(128, dropout=0.2, recurrent_dropout=0.2),
    LSTMCell(128, dropout=0.2, recurrent_dropout=0.2),
]

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(RNN(cells))
model.add(Dense(1, activation='sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=5,
          validation_data=(x_test, y_test))
score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

Loading data...
25000 train sequences
25000 test sequences
Pad sequences (samples x time)
x_train shape: (25000, 80)
x_test shape: (25000, 80)
New stacked LSTM: 30s/epoch on CPU (15pct faster)
Build model...
Train...
Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 330s 13ms/step - loss: 0.4524 - acc: 0.7876 - val_loss: 0.3879 - val_acc: 0.8293
Epoch 2/5
25000/25000 [==============================] - 310s 12ms/step - loss: 0.2929 - acc: 0.8821 - val_loss: 0.3775 - val_acc: 0.8322
Epoch 3/5
25000/25000 [==============================] - 312s 12ms/step - loss: 0.2153 - acc: 0.9182 - val_loss: 0.4019 - val_acc: 0.8318
Epoch 4/5
25000/25000 [==============================] - 313s 13ms/step - loss: 0.1564 - acc: 0.9428 - val_loss: 0.4324 - val_acc: 0.8256
Epoch 5/5
25000/25000 [==============================] - 77s 3ms/step
Test score: 0.5160929471731186
Test accuracy: 0.82476


# CuDNNGRU

In [15]:
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import CuDNNGRU
from keras.datasets import imdb

max_features = 20000
# cut texts after this number of words (among top max_features most common words)
maxlen = 80
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(CuDNNGRU(128))
model.add(Dense(1, activation='sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=5,
          validation_data=(x_test, y_test))
score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

Loading data...
25000 train sequences
25000 test sequences
Pad sequences (samples x time)
x_train shape: (25000, 80)
x_test shape: (25000, 80)
Build model...
Train...
Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 19s 744us/step - loss: 0.4210 - acc: 0.7992 - val_loss: 0.3436 - val_acc: 0.8478
Epoch 2/5
25000/25000 [==============================] - 16s 647us/step - loss: 0.2289 - acc: 0.9090 - val_loss: 0.3763 - val_acc: 0.8298
Epoch 3/5
25000/25000 [==============================] - 16s 635us/step - loss: 0.1208 - acc: 0.9559 - val_loss: 0.4861 - val_acc: 0.8285
Epoch 4/5
25000/25000 [==============================] - 16s 632us/step - loss: 0.0585 - acc: 0.9801 - val_loss: 0.5551 - val_acc: 0.8285
Epoch 5/5
25000/25000 [==============================] - 5s 198us/step
Test score: 0.7516524196100235
Test accuracy: 0.82092


# CuDNNLSTM

In [14]:
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import CuDNNLSTM
from keras.datasets import imdb

max_features = 20000
# cut texts after this number of words (among top max_features most common words)
maxlen = 80
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(CuDNNLSTM(128))
model.add(Dense(1, activation='sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=5,
          validation_data=(x_test, y_test))
score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

Loading data...
25000 train sequences
25000 test sequences
Pad sequences (samples x time)
x_train shape: (25000, 80)
x_test shape: (25000, 80)
Build model...
Train...
Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 19s 767us/step - loss: 0.4354 - acc: 0.7949 - val_loss: 0.3924 - val_acc: 0.8274
Epoch 2/5
25000/25000 [==============================] - 16s 656us/step - loss: 0.2558 - acc: 0.8984 - val_loss: 0.3732 - val_acc: 0.8390
Epoch 3/5
25000/25000 [==============================] - 16s 636us/step - loss: 0.1590 - acc: 0.9393 - val_loss: 0.4327 - val_acc: 0.8279
Epoch 4/5
25000/25000 [==============================] - 16s 642us/step - loss: 0.0981 - acc: 0.9642 - val_loss: 0.5334 - val_acc: 0.8214
Epoch 5/5
25000/25000 [==============================] - 5s 211us/step
Test score: 0.656510255368352
Test accuracy: 0.81992


# ロイター

In [17]:
from __future__ import print_function

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import RNN, SimpleRNNCell
from keras.datasets import reuters
from keras.utils import np_utils

max_features = 20000
# cut texts after this number of words (among top max_features most common words)
maxlen = 80
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=max_features)
y_train = np_utils.to_categorical(y_train)
y_test = np_utils.to_categorical(y_test)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('New stacked LSTM: 30s/epoch on CPU (15pct faster)')
cells = [
    SimpleRNNCell(128, dropout=0.2, recurrent_dropout=0.2),
]

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(RNN(cells))
model.add(Dense(46, activation='sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=20,
          validation_data=(x_test, y_test))
score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

Loading data...
8982 train sequences
2246 test sequences
Pad sequences (samples x time)
x_train shape: (8982, 80)
x_test shape: (2246, 80)
New stacked LSTM: 30s/epoch on CPU (15pct faster)
Build model...
Train...
Train on 8982 samples, validate on 2246 samples
Epoch 1/20
8982/8982 [==============================] - 16s 2ms/step - loss: 2.6442 - acc: 0.3213 - val_loss: 2.4580 - val_acc: 0.3620
Epoch 2/20
8982/8982 [==============================] - 13s 1ms/step - loss: 2.4320 - acc: 0.3517 - val_loss: 2.4306 - val_acc: 0.3620
Epoch 3/20
8982/8982 [==============================] - 14s 2ms/step - loss: 2.4301 - acc: 0.3517 - val_loss: 2.4395 - val_acc: 0.3620
Epoch 4/20
8982/8982 [==============================] - 13s 1ms/step - loss: 2.4245 - acc: 0.3517 - val_loss: 2.4418 - val_acc: 0.3620
Epoch 5/20
8982/8982 [==============================] - 13s 1ms/step - loss: 2.3763 - acc: 0.3536 - val_loss: 2.3305 - val_acc: 0.3776
Epoch 6/20
8982/8982 [==============================] - 13s 1ms/

# 各モデルの結果を比較
|Model |Train time|Predict time|Test score  |Test accuracy  |
|---|---|---|---|---|
|SimpleRNN |44s |11s |0.52080 |0.75712 |
|SimpleRNNCell |44s |11s |0.64211 |0.7148 |
|LSTM |168s |43s |0.58744 |0.81792 |
|GRU |130s |33s |0.63379 |0.8212 |
|GRUCell |134s |31s |0.64971 |0.8196 |
|CuDNNGRU |16s |5s |0.75165 |0.82092 |
|CuDNNLSTM |16s |5s |0.65651 |0.81992 |

以上の結果より、処理速度が最も早いのは「CuDNNGRU」、「CuDNNLSTM」である。他のモデルよりも早いのはGPUを使用したモデルのため、他よりも圧倒的に処理速度が向上している。精度に関しては「SimpleRNN」以外は同程度の精度になっている。以上の結果より、CPUのみ環境で処理速度を重視するのであれば、「SimpleRNN」。処理速度よりも精度であれば、「LSTM」、「GRU」が最適だと思われる。また、GPU環境であれば、「CuDNNGRU」、「CuDNNLSTM」を使用するのが良い。

# 説明

## LSTMやGRUにはゲートと呼ばれるものがあるがこれにはどのような働きがあるか
ゲートには output ゲート、forget ゲート、input ゲートの3つある。output ゲートは次の隠れ層への出力を制御する。forget ゲートは記憶セルからの情報をどれだけ消すか(忘れる)を制御する。input ゲートは新しい記憶を記憶セルに保存する際に、どの情報を保存するかを制御する。

## SimpleRNN、LSTM、GRUはどのように使い分ければ良いか
SimpleRNNはSimpleRNN、LSTM、GRUの三つの中で最もパラメータが数が少ないモデルである。そのため、他の二つよりも計算コストが少なく処理も早い。しかい、長い時系列データを学習すると勾配消失が起きやすいため、大規模データには向かいない。よって、時系列データが短く、処理を早くしたい場合に用いるのが良い。<br>
LSTMはSimpleRNNの欠点であった長い時系列データでの勾配消失を改良したモデルである。三つの中では最もモデルの表現力が高い。しかし、パラメータが多いため、処理が遅い。よって、時系列データが長く、モデルの表現力を上げたい場合に用いるのが良い。<br>
GRUはLSTMよりパラメータ数を削減したモデル。LSTMの改良モデルのため、長い時系列データにも対応が可能である。また、LSTMに比べパラメータが少ないので計算コストが少なく処理も早い。しかし、パラメータを削減しているため、LSTMと比べて表現力が低い。よって、LSTMほど表現力が必要なく、処理を高速化したい場合に用いるのが良い。

## StackedRNNを使うことはどういった効果があるか
StackedRNN を使用すると RNNレイヤー深く重ねる(層を深く)できる。層を深くすることで表現力の高いモデルを作る事ができる。

## ConvLSTMはどういった場面で使われるか
LSTMで2階テンソルを扱えるので、動画処理で利用が可能となる。